### PR TITLE
Update StringBuilder documentation to be consistent about internal buffer copying

### LINKED ISSUE #47735: Inconsistent/contradictory statements on StringBuilder marshaling in PInvoke
There are a few places in the docs that seem to state inconsistent/contradictory facts on how `StringBuilder` is treated during PInvoke marshaling. Please clarify which is correct, and adjust the docs as appropriate to reduce future confusion.

### Some docs state that when a `StringBuilder` is passed by value to PInvoke, a pointer to the internal buffer is passed to native, and _no copy occurs_. E.g.:

* Docs >.NET > .NET Framework > Interoperate with unmanaged code > Copying and Pinning:
sub-section "[System.String and System.Text.StringBuilder](https://docs.microsoft.com/en-us/dotnet/framework/interop/copying-and-pinning#systemstring-and-systemtextstringbuilder)": 

> When a System.Text.StringBuilder is passed by value, the marshaler passes a reference to the internal buffer of the StringBuilder directly to the caller. The caller and callee must agree on the size of the buffer. The caller is responsible for creating a StringBuilder of adequate length. The callee must take the necessary precautions to ensure that the buffer is not overrun. StringBuilder is an exception to the rule that reference types passed by value are passed as In parameters by default. It is always passed as In/Out.

* Docs > .NET > .NET Framework > Interoperate with unmanaged code > Default Marshaling for Strings
sub-section "[Fixed-Length String Buffers](https://docs.microsoft.com/en-us/dotnet/framework/interop/default-marshaling-for-strings#fixed-length-string-buffers)":

> [...] The solution is to pass a StringBuilder buffer as the argument instead of a String. A StringBuilder can be dereferenced and modified by the callee, provided it does not exceed the capacity of the StringBuilder. It can also be initialized to a fixed length. [...]

### Conversely, other docs explicitly recommend avoiding the use of `StringBuilder` stating that a _copy of the internal buffer always occurs_. E.g.:

* Docs > .NET > .NET fundamentals > Native interoperability best practices
sub-section "[String parameters](https://docs.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#string-parameters)":
> AVOID StringBuilder parameters. StringBuilder marshaling always creates a native buffer copy.  [...]

* Docs > .NET > .NET fundamentals > CA1838: Avoid StringBuilder parameters for P/Invokes
[entire page](https://docs.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1838):

> Marshaling of StringBuilder always creates a native buffer copy, resulting in multiple allocations for one P/Invoke call. To marshal a StringBuilder as a P/Invoke parameter, the runtime will:
    - Allocate a native buffer.
    - If it is an In parameter, copy the contents of the StringBuilder to the native buffer.
    - If it is an Out parameter, copy the native buffer into a newly allocated managed array.
[...]

### PR DESCRIPTION
**Pull request by jkoritzinsky**
## Summary

StringBuilder marshalling always copies the buffer to ensure that there is a null terminator at the end. It never reuses the internal buffer directly.

Fixes dotnet/runtime#47735

### CODE REVIEW

File: docs/framework/interop/copying-and-pinning.md [hunk 1]
```diff
@@ -71,7 +71,7 @@ As an optimization when either string type is marshaled by value (such as a Unic
...
 
-When a <xref:System.Text.StringBuilder?displayProperty=nameWithType> is passed by value, the marshaler passes a reference to the internal buffer of the **StringBuilder** directly to the caller. The caller and callee must agree on the size of the buffer. The caller is responsible for creating a **StringBuilder** of adequate length. The callee must take the necessary precautions to ensure that the buffer is not overrun. **StringBuilder** is an exception to the rule that reference types passed by value are passed as In parameters by default. It is always passed as In/Out.
+When a <xref:System.Text.StringBuilder?displayProperty=nameWithType> is passed by value, the marshaler passes a reference to a temporary copy of the internal buffer of the **StringBuilder** directly to the caller. The caller and callee must agree on the size of the buffer. The caller is responsible for creating a **StringBuilder** of adequate length. The callee must take the necessary precautions to ensure that the buffer is not overrun. **StringBuilder** is an exception to the rule that reference types passed by value are passed as In parameters by default. It is always passed as In/Out.
```
Comment (elinor-fung) [inline docs/framework/interop/copying-and-pinning.md:74]:
```suggestion
When a <xref:System.Text.StringBuilder?displayProperty=nameWithType> is passed by value, the marshaler passes a reference to a temporary copy of the internal buffer of the **StringBuilder** to the caller. The caller and callee must agree on the size of the buffer. The caller is responsible for creating a **StringBuilder** of adequate length. The callee must take the necessary precautions to ensure that the buffer is not overrun. **StringBuilder** is an exception to the rule that reference types passed by value are passed as In parameters by default. It is always passed as In/Out.
```

**Review by elinor-fung — COMMENTED** [review]
Should we also clarify the wording in the other doc the issue pointed at (https://docs.microsoft.com/dotnet/framework/interop/default-marshaling-for-strings#fixed-length-string-buffers: 'A `StringBuilder` can be dereferenced and modified by the callee') to indicate that it is a copy of the internal buffer?

**Review by elinor-fung — APPROVED** [review]

**Comment by adegeo** [comment]
@elinor-fung @jkoritzinsky Just to doublecheck, this content was written for .NET Framework, but I assume it applies to both .NET and .NET Framework?

**Comment by jkoritzinsky** [comment]
Yes this applies to both. We have some other documentation in the shared folder that’s already been updated.

**Pull request merged**